Multiple fixes (code logic and issues)

AntonioDePau opened the pull request.

AntonioDePau committed:
Fixes and improvements

- Fixed window size
- Added precize zoom with CTRL pressed while scrolling
- Re-added the possibility to load an image as an argument passed to the .exe

AntonioDePau committed:
Image as an argument re-added

Not sure if this is how I should re-implement this, but here goes...

AntonioDePau committed:
Fixed indentation

AntonioDePau committed:
Fixed indent (again)

AntonioDePau committed:
Fixed Zoom Level label position

AntonioDePau committed:
File count must be written as a short

AntonioDePau committed:
Removed extra space added previously

I keep messing formatting up...

AntonioDePau committed:
Removed yet another extra space

AntonioDePau commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -182,10 +182,10 @@ public class TextureConverter {
             file.RemoveRange(8, 2);
-            file.InsertRange(8, BitConverter.GetBytes(header.FileCount).ToList());
+            file.InsertRange(8, BitConverter.GetBytes((short)header.FileCount).ToList());
 
             file.RemoveRange(10, 2);
-            file.InsertRange(10, BitConverter.GetBytes(header.FileCountB).ToList());
+            file.InsertRange(10, BitConverter.GetBytes((short)header.FileCountB).ToList());
```
FileCount and FileCountB must be written as short, otherwise the whole file will be 4 bytes longer than expected.

Andrew9x replied:
See my comment above.

AntonioDePau committed:
Added log regarding colors to help debugging

A log file will be created in the folder where the original PNG files were exported for each PNG file repacked into a XET file if the original XET file is loaded in the software at the time of repacking.

AntonioDePau committed:
Added more info to the log

AntonioDePau committed:
Added palette colors to Colors property

This will help for the logs generated when repacking

AntonioDePau committed:
Added unindexed image creation

Implemented this to see if colors could have been lost during the indexation process.
Results are identical, issue [#5](https://github.com/AntonioDePau/psp_xettex_viewer/issues/5) remains open.

Andrew9x commented:
Give the pull request a title please.

Andrew9x commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -75,8 +75,39 @@ public class TextureConverter {
+            int row = 0;
+            int bitsPerPixelMultiplier = 8 / texture.BitsPerPixel;
+
+            List<System.Drawing.Color> colors = new List<System.Drawing.Color>();
```
Move this to the imports. Only use Color class here.

Andrew9x commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -75,8 +75,39 @@ public class TextureConverter {
+
+            List<System.Drawing.Color> colors = new List<System.Drawing.Color>();
+            
+            for (int c = 0; c < texture.Palette.Length / 4; c++) {
```
Move this to a private method, with return type: List<Colors>.
As parameter pass the texture.Palette array.

On this level call this only:
texture.colors.AddAll( privateMethod );

Andrew9x commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -75,8 +75,39 @@ public class TextureConverter {
+                texture.Colors.Add(col);
+            }
+
+            for (int x = 0; x < texture.Unswizzled.Length; x++) {
```
Move this to a private method., with return type of Bitmap.
Pass the color array, the unswizzled array and the with as parameter.

AntonioDePau replied:
Isn't this what the method GetBitmapFromPalette is already doing?
Also, not sure I committed this already, but there are now 2 methods: GetIndexedBitmapFromPalette and GetBitmapFromPalette. Each need a different way of iterating through Unswizzled. So, I would rather keep those two methods as they are.

Andrew9x commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -75,8 +75,39 @@ public class TextureConverter {
+            }
+
+            for (int x = 0; x < texture.Unswizzled.Length; x++) {
+                if (x >= texture.Width + (texture.Width * row)) row++;
```
Alawys use brackets.

AntonioDePau committed:
Re-added 8bpp support for non-indexed bitmap

Andrew9x commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -182,10 +214,10 @@ public class TextureConverter {
 
             file.RemoveRange(8, 2);
-            file.InsertRange(8, BitConverter.GetBytes(header.FileCount).ToList());
+            file.InsertRange(8, BitConverter.GetBytes((short)header.FileCount).ToList());
```
Instead of casting change these fields in TexHeaderMap from integer to short datatype.

AntonioDePau replied:
Thought you changed them and that's what they were originally :D alright, will change.

Andrew9x replied:
Yes it does ring a bell. I didn't understand the purpose back then.

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -12,17 +12,20 @@ namespace ConsoleProject.View {
 
     public class Frontend {
 
+        private string AppName = "XetTex Viewer";
```
This can be static. Static variable names should be uppercase.

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -108,7 +104,12 @@ public class Frontend {
 
-            infoPanel.Controls.Add(labelInfo);
+            infoPanel.Controls.Add(labelImageInfo);
+            //Not sure if the following 3 lines should be here or somewhere else?
```
They are fine here currently. Will be moved to a private method later.

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -156,18 +178,20 @@ public class Frontend {
                 imageListBox.SelectedIndex = 0;
             }
             DisplayImage();
+            form.Text = AppName + " - " + Path.GetFileName(filename);
```
I don't see why is this info required.

AntonioDePau replied:
Just for conveniency of knowing which file is currently being loaded.
Helped me a couple of times when trying multiple repackings with the same images and a repack wouldn't correctly load (without displaying any errors). But I understand where this is coming from, and there should be an error message displayed at some point rather than doing this.

However, most apps that open files do display the filename in the title bar (notepad, paint, etc.)

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -185,22 +209,55 @@ public class Frontend {
                                 if (palette.IndexOf(c) == -1) {
                                     if (palette.Count < 256) palette.Add(c);
                                 }
+                                if (completeColors.IndexOf(c) == -1) completeColors.Add(c);
```
Add missing brackets.

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -185,22 +209,55 @@ public class Frontend {
 
                         palette.Sort((x, y) => x.A.CompareTo(y.A));
+                        
+                        if (log) {
```
We'll use a logging framework for this. Please add a TODO comment here: "replace with Logger"

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -241,6 +302,7 @@ public class Frontend {
                                 dataIndex++;
                             }
                         }
+                        if (log) File.WriteAllText(file + ".log.txt", logContent);
```
Add missing brackets.

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -125,13 +126,34 @@ public class Frontend {
+        }
 
         private void DrawPicture(MouseEventArgs mouseEvent = null) {
+            if (obmp == null) return;
```
Add missing brackets. Also write an info log to the console, why is this an error. If it's an error maybe throwing an exception is better. I haven't yet understood the connection between bmp and obmp.

AntonioDePau replied:
I think this happens when no XET (and thus bitmap) has been loaded.
Example: You start the EXE without providing a file, so the form shows, but no default image is loaded.
If at this point you do not return from this function, you get an exception when trying to zoom in or out.

As for bmp vs obmp, bmp is the currently displayed image (including zoom level) while obmp is the original image (default zoom) which allows for the ratio calculation.

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -125,13 +126,34 @@ public class Frontend {
             return listBox;
         }
+        
+        private void ParseNewFile(string[] fileList) {
```
This method should not be in the FrontEnd layer. A better place would be in the TextureConverter class.

AntonioDePau replied:
Moving it to TextureConverter makes it more difficult to interact with Frontend and populate the "images" list only if the file is valid though. I'll see what I can do...

AntonioDePau replied:
So... I moved it here (TextureConverter.cs):
![image](https://user-images.githubusercontent.com/23122727/93709020-d2b16880-fb32-11ea-8da0-42e68d23be5b.png)

However, since I cannot interact with the Frontend, the conditions have to be based on the return value of the method, so it means that I have double (triple?) the code in Frontend to achieve the same result as before:
![image](https://user-images.githubusercontent.com/23122727/93709036-faa0cc00-fb32-11ea-8ebe-7f27e97f8b3d.png)

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -125,13 +126,34 @@ public class Frontend {
         }
+        
+        private void ParseNewFile(string[] fileList) {
+            if (fileList.Length > 0) { // is this IF required?
```
I'm surprised that this doesn't throw a null pointer exception.

AntonioDePau replied:
Not sure what this means or what I should do about it.
This is just to check if there's any file at all being passed to this method.
I could also return if the length is Zero if you prefer.

Andrew9x replied:
Check ShowForm signature: arguments is an optional parameter. If it is not provided then it'll get initialized as null.

Here you are calling a property on a null object. Which should throw an exception.

Andrew9x commented on ConsoleProject/Application.cs:
```diff
@@ -6,8 +6,8 @@ namespace ConsoleProject {
-        public static void Main() {
-            new Frontend().ShowForm();
+        public static void Main(string[] arguments) {
+            new Frontend().ShowForm(arguments);
```
I don't like that we are passing "arguments" to the frontend.
Renaming them to "filelist" would be better in the ShowForm method signature.

However the ultimate goal is to seperate the console and the frontend layers. Thus the frontEnd should not require the console inputs. The underlying layers (eg texture converter) should be able to service both the windows form frontend and the console frontend.

AntonioDePau replied:
If this is only a naming issue, fine, it can be renamed.
However, I would prefer to keep the possibility to pass files as arguments, not necessarily via console input, but rather as a drag-and-drop on the exe or double-click with the exe set as the default opening program.

Andrew9x replied:
"double-click with the exe set as the default opening program." 
Haven't thought about this, good point. 

At the Application layer we can work with it as arguments, thats fine.
But the form should already know what is it getting as parameter, thus the need of rename.

Some sort of validation might be a good idea too. Perhaps add a todo here, to validate the input arguments.

AntonioDePau committed:
Updated according to pull request comments

Renamed arguments to fileList
Added TODO

AntonioDePau committed:
Updated according to pull request comments

- Moved ParseNewFile to TextureConverter (and adjusted code to react to the return value)
- Added missing brackets

AntonioDePau committed:
Updated according to pull request comments

- Added missing brackets
- Removed casting
- Moved "ParseNewFile" here
- Added TODO for logging
- Moved the palette parsing to its own method

AntonioDePau committed:
Updated according to pull request comments

- Changed int to short for file count in order to not have to cast it into short later

AntonioDePau committed:
Re-fixed iteration through Width and Height

Don't know how this keeps happening...
Should fix https://github.com/AntonioDePau/psp_xettex_viewer/issues/5

Andrew9x commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -75,20 +88,61 @@ public class TextureConverter {
+        
+        private List<Color> GetColorsFromPalette(byte[] palette) {
+            List<Color> colors = new List<Color>();
+            for (int c = 0; c < palette.Length / 4; c++) {
```
Don't divide the length by 4.
Increment by using `c+=4` instead of `c++`;
Also use`i` as index variable instead of `c`.

Benefits: we don't need the multiplication of `c * 4`.

AntonioDePau replied:
Done!

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -146,7 +170,7 @@ public class Frontend {
         }
 
-        private void LoadImages() {
+        public void LoadImages(string filename) {
```
Why did this method had to become public?

AntonioDePau replied:
Mhhh... Forgot to revert it back to Private.
That was when I thought I could interact with it from TextureConverter. Fixed.

AntonioDePau committed:
Changed "public" back to "private"

AntonioDePau committed:
Updated GetColorsFromPalette

- Changed "c" to "i"
- Increment by 4 instead of 1, not having to divide by 4

Andrew9x commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -30,6 +30,19 @@ public class TextureConverter {
+                    return ParseFile(fileList[0]);
+                } catch (InvalidDataException ex) {
+                    Console.WriteLine(ex);
+                    return null;
```
Avoid returning with null. Returning with null always causes a null check in the caller layer. Usually only the not-null case is handled, without the else branch. 

I changed the original null return to exception throwing. An exception must be handled, unlike a null return.

Andrew9x replied:
Here I would remove the try catch, and move it to the caller layer. Let the exception propagate upward.

Andrew9x commented on ConsoleProject/src/services/TextureConverter.cs:
```diff
@@ -99,7 +153,9 @@ public class TextureConverter {
 
             for (int x = 0; x < texture.Unswizzled.Length; x++) {
-                if (x >= texture.Width + (texture.Width * row)) row++;
+                if (x >= texture.Width + (texture.Width * row)){
```
Could you explain this condition? It feels like it can be replaced with modulo arithmetic.
if ( x % texture.Width == 0) { row++; }

AntonioDePau replied:
Indeed, changed to modulo!

Andrew9x commented on ConsoleProject/src/view/Frontend.cs:
```diff
@@ -46,18 +49,23 @@ public class Frontend {
-                    }
+                string[] filesDropped = (string[])e.Data.GetData(DataFormats.FileDrop, false);
+                List<Texture> listDropped = new TextureConverter().ParseNewFile(filesDropped);
+                if (listDropped != null) {
```
Replace null check with exception handling. Also extract the common part to a private function here. This will solve your code duplication concerns too.

AntonioDePau committed:
Removed "ParseNewFile"

AntonioDePau committed:
Added "TryTextureFile"

- This is were exceptions when trying to read the texture will be caught

AntonioDePau committed:
Added TODO

- Added Todo: handle multiple files
